### PR TITLE
fix(RHTAPBUGS-560): add subdirectory parameter to tasks

### PR DESCRIPTION
**Pull request by mmalina**
This is to avoid multiple PRs running in parallel overriding each other's data.

This updates the tasks to be able to use a subdirectory inside a workspace.

There will be another PR to update the pipelines to use these.

### CODE REVIEW

**Comment by johnbieren** [comment]
> > We already have `*Path` parameters for every file in the workspace. Why do we need to add a subdirectory parameter?
> 
> @johnbieren which tasks are you referring to? I'm sure not all.
> 
> These are the tasks I'm changing:
> 
> ```
> apply-mapping/0.6
> collect-data/0.2
> create-internal-request/0.7
> create-pyxis-image/0.5
> push-sbom-to-pyxis/0.3
> push-snapshot/0.11
> ```
> 
> `collect-data` doesn't have any parameters to specify where to save data, so it's definitely needed there.
> 
> `apply-mapping` does not allow to specify where to store the mapped file - it's hardcoded to `mapped_snapshot.json`. I could add another parameter for that if you think that's better.
> 
> `create-internal-request` doesn't have any such parameters. This is what it does:
> 
> ```shell
> IR_DIR="$(workspaces.input.path)/$(params.subdirectory)/internal-request"
> IR="${IR_DIR}/ir-$(params.pipelineRunName)-$(context.taskRun.uid).yaml"
> ```
> 
> (I added the subdir.) So maybe here it's not really needed. What's really wrong there is the fact that it saves the absolute path of a file and saves it as a result for other tasks to use - I guess it works, but you should never rely on the workspace mountpoints being exactly the same across tasks. I decided to leave it. Let me know what you think I should do here. I can surely revert to the original.
> 
> `create-pyxis-image`, `push-sbom-to-pyxis` and `push-snapshot` - you're right here the path could be provided as a parameter. I guess maybe I wanted to be consistent with the others and also `snapshotPath` can stay on the default. But yeah, now that I think of it, it's probably better to just use the parameter.

Well, `collect-data` is the first one to run. If we need to use subdirectories because of a bug, it should automatically use a subdirectory without a parameter needing to be passed IMO. The path to the file could be a task result

`apply-mapping` could have some logic added to put the `mapped_snapshot` in the same directory as the `snapshotPath` and output the path to it as a result or something.

I don't know much about `create-internal-request`

For the `pyxis` tasks and `push-snapshot` we already have `snapshotPath`. I really don't think having a subdirectory parameter too is a good idea. We don't need two parameters to get one file.

I am basically saying I am -1 to the idea that we should have two parameters per task to get a single file. That should not be needed. If the *Path parameters need to change, sure. Let's not do two parameters for a path to one file

**Comment by mmalina** [comment]
I reverted to the original versions for `create-pyxis-image`, `push-sbom-to-pyxis` and `push-snapshot`. So that leaves us with 3 changed tasks.

> Well, `collect-data` is the first one to run. If we need to use subdirectories because of a bug, it should automatically use a subdirectory without a parameter needing to be passed IMO. The path to the file could be a task result

You're right that currently collect-data is always the first one, but that can always change. Personally I'd prefer to define the subdir from the pipeline and not the task. Also, a similar approach is already used in the `clone-config-file` task. Would that be acceptable for you?

> `apply-mapping` could have some logic added to put the `mapped_snapshot` in the same directory as the `snapshotPath` and output the path to it as a result or something.

Hmm, looking at `apply-mapping`, I wonder if we even need the mapped_snapshot.json file? Why don't we simply override the original snapshot file? It's not like we need to use the original somewhere else, or do we? If we do need it, yeah, I guess it could just use the same directory as the snapshot. That way there would be no need for an extra param or result. WDYT?

> If the pipelines are buggy without using a subdirectory, don't make it optional IMO

It's not optional. All the parameters I added are required.

**Comment by johnbieren** [comment]
> I reverted to the original versions for `create-pyxis-image`, `push-sbom-to-pyxis` and `push-snapshot`. So that leaves us with 3 changed tasks.
> 
> > Well, `collect-data` is the first one to run. If we need to use subdirectories because of a bug, it should automatically use a subdirectory without a parameter needing to be passed IMO. The path to the file could be a task result
> 
> You're right that currently collect-data is always the first one, but that can always change. Personally I'd prefer to define the subdir from the pipeline and not the task. Also, a similar approach is already used in the `clone-config-file` task. Would that be acceptable for you?
Well, collect-data kind of always has to be first. The vision is that we don't really pass many (or any at all) parameters to the pipeline and instead get them all from CRs. We don't have access to CR data until collect-data runs. `clone-config-file` is set to be deprecated afaik as we won't be using external config files. I can see the value in wanting to define the subdir at the pipeline level though
> 
> > `apply-mapping` could have some logic added to put the `mapped_snapshot` in the same directory as the `snapshotPath` and output the path to it as a result or something.
> 
> Hmm, looking at `apply-mapping`, I wonder if we even need the mapped_snapshot.json file? Why don't we simply override the original snapshot file? It's not like we need to use the original somewhere else, or do we? If we do need it, yeah, I guess it could just use the same directory as the snapshot. That way there would be no need for an extra param or result. WDYT?
I can't think of any reason we can't overwrite the existing snapshot file
> 
> > If the pipelines are buggy without using a subdirectory, don't make it optional IMO
> 
> It's not optional. All the parameters I added are required.
Yeah, optional was the wrong word choice by me. I meant configurable. Force it by putting it hardcoded in the task if it is required IMO

File: catalog/task/collect-data/0.2/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+## Parameters
+
+| Name                 | Description                                        | Optional | Default value |
+|----------------------|----------------------------------------------------|----------|---------------|
+| release              | Namespaced name of the Release                     | No       | -             |
+| releaseplan          | Namespaced name of the ReleasePlan                 | No       | -             |
+| releaseplanadmission | Namespaced name of the ReleasePlanAdmission        | No       | -             |
+| releasestrategy      | Namespaced name of the ReleaseStrategy             | No       | -             |
+| snapshot             | Namespaced name of the Snapshot                    | No       | -             |
+| subdirectory      | Subdirectory inside the workspace to be used. | Yes       | -             |
```
Comment (johnbieren) [inline catalog/task/collect-data/0.2/README.md:19]: can you fix the spacing to be like the others?

File: catalog/task/collect-data/0.2/collect-data.yaml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        - name: "SNAPSHOT"
+          value: '$(params.snapshot)'
+        - name: "SUBDIR"
+          value: '$(workspaces.data.path)/$(params.subdirectory)'
+      script: |
+        #!/usr/bin/env sh
+        set -e
+
+        if [ -n "$(params.subdirectory)" ]; then
+          mkdir -p $SUBDIR
```
Comment (johnbieren) [inline catalog/task/collect-data/0.2/collect-data.yaml:61]: could probably just `cd` in after this and not have to write it on every line after, but fine either way

**Comment by johnbieren** [comment]
One of your tests has been in progress for 4 hours so I think there is an issue (canceling the workflow will probably print which one it was)

File: catalog/task/collect-data/0.2/README.md (hunk 1 above)
Comment (mmalina) [reply catalog/task/collect-data/0.2/README.md]: Fixed.

File: catalog/task/collect-data/0.2/collect-data.yaml (hunk 2 above)
Comment (mmalina) [reply catalog/task/collect-data/0.2/collect-data.yaml:61]: Good idea, done.

**Review by johnbieren — APPROVED** [review]

**Comment by mmalina** [comment]
> One of your tests has been in progress for 4 hours so I think there is an issue (canceling the workflow will probably print which one it was)

It's really hard to say what happened, but I don't think it's related to this PR. When I cancelled it, the last lines were:
```
[cleanup : delete-crs] + kubectl delete release release-sample
[cleanup : delete-crs] release.appstudio.redhat.com "release-sample" deleted
[cleanup : delete-crs] + kubectl delete releaseplan releaseplan-sample
[cleanup : delete-crs] releaseplan.appstudio.redhat.com "releaseplan-sample" deleted
[cleanup : delete-crs] + kubectl delete releaseplanadmission releaseplanadmission-sample
[cleanup : delete-crs] releaseplanadmission.appstudio.redhat.com "releaseplanadmission-sample" deleted
[cleanup : delete-crs] + kubectl delete releasestrategy releasestrategy-sample
[cleanup : delete-crs] releasestrategy.appstudio.redhat.com "releasestrategy-sample" deleted
[cleanup : delete-crs] + kubectl delete snapshot snapshot-sample
[cleanup : delete-crs] snapshot.appstudio.redhat.com "snapshot-sample" deleted
```
I've seen a similar thing in the past locally and I believe where it gets sometimes stuck is on tkn: `tkn pr logs $PIPELINERUN -f` - most of the time it works as expected, but sometimes it can get stuck even after the pr completes.

I tried it again locally and it works for me. But on github, the very next run got stuck again.

I refactored it a bit to avoid this.

File: .github/scripts/test_tekton_tasks.sh
Comment (johnbieren) [inline .github/scripts/test_tekton_tasks.sh:1]: hate to be a pain, but this should really be its own commit/pr

Comment (mmalina) [reply .github/scripts/test_tekton_tasks.sh:1]: ok, here: https://github.com/redhat-appstudio/release-service-bundles/pull/177

Comment (mmalina) [reply .github/scripts/test_tekton_tasks.sh:1]: it's merged and this PR is rebased. it should be ready to go hopefully.

**Review by johnbieren — APPROVED** [review]
